File: native/macos/SkillFlow/Sources/Daemon/DaemonClient.swift
import Foundation

struct DaemonClient {
    private let endpointPath: String
    private let session: URLSession

    init(
        endpointPath: String = DaemonClient.defaultEndpointPath(),
        session: URLSession = .shared
    ) {
        self.endpointPath = endpointPath
        self.session = session
    }

    static func defaultEndpointPath() -> String {
        let applicationSupport = URL(
            fileURLWithPath: NSHomeDirectory(),
            isDirectory: true
        )
        .appendingPathComponent("Library/Application Support", isDirectory: true)
        .appendingPathComponent("SkillFlow", isDirectory: true)
        .appendingPathComponent("runtime", isDirectory: true)
        .appendingPathComponent("daemon-service.json")

        return applicationSupport.path
    }

    func invoke<Result: Decodable, Parameters: Encodable>(
        _ method: String,
        parameters: Parameters? = nil,
        requestID: String = UUID().uuidString
    ) async throws -> Result {
        let endpoint = try loadEndpoint()
        guard let url = URL(string: "http://\(endpoint.address)/invoke") else {
            throw DaemonClientError.invalidEndpoint
        }

        let nativeRequest = NativeAPIRequest(
            version: NativeAPIVersion.current,
            method: method,
            parameters: parameters,
            requestID: requestID
        )
        let serviceRequest = DaemonServiceRequest(
            method: "native.api",
            parameters: nativeRequest
        )

        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "POST"
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        urlRequest.setValue(endpoint.token, forHTTPHeaderField: "X-SkillFlow-Token")
        urlRequest.httpBody = try JSONEncoder().encode(serviceRequest)

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: urlRequest)
        } catch {
            throw DaemonClientError.transport(error.localizedDescription)
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw DaemonClientError.invalidResponse
        }
        if httpResponse.statusCode == HTTPStatus.unauthorized {
            throw DaemonClientError.unauthorized
        }
        guard (200..<300).contains(httpResponse.statusCode) else {
            throw DaemonClientError.transport("HTTP \(httpResponse.statusCode)")
        }

        let decoded: DaemonServiceResponse<NativeAPIResponse<Result>>
        do {
            decoded = try JSONDecoder().decode(
                DaemonServiceResponse<NativeAPIResponse<Result>>.self,
                from: data
            )
        } catch {
            throw DaemonClientError.invalidResponse
        }

        guard decoded.ok else {
            throw DaemonClientError.service(decoded.error ?? "unknown service error")
        }

        guard let nativeResponse = decoded.result else {
            throw DaemonClientError.invalidResponse
        }
        guard nativeResponse.ok else {
            let apiError = nativeResponse.error
            throw DaemonClientError.api(
                code: apiError?.code ?? "internal_error",
                messageKey: apiError?.messageKey,
                message: apiError?.message
            )
        }
        guard let result = nativeResponse.result else {
            throw DaemonClientError.missingResult
        }
        return result
    }

    func invoke<Result: Decodable>(
        _ method: String,
        requestID: String = UUID().uuidString
    ) async throws -> Result {
        try await invoke(method, parameters: Optional<NoParameters>.none, requestID: requestID)
    }

    private func loadEndpoint() throws -> DaemonEndpoint {
        let endpointURL = URL(fileURLWithPath: endpointPath)
        guard FileManager.default.fileExists(atPath: endpointURL.path) else {
            throw DaemonClientError.endpointUnavailable
        }

        do {
            let endpoint = try JSONDecoder().decode(
                DaemonEndpoint.self,
                from: Data(contentsOf: endpointURL)
            )
            guard !endpoint.address.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
                  !endpoint.token.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
                throw DaemonClientError.invalidEndpoint
            }
            return endpoint
        } catch let error as DaemonClientError {
            throw error
        } catch {
            throw DaemonClientError.invalidEndpoint
        }
    }
}

private enum HTTPStatus {
    static let unauthorized = 401
}

private struct NoParameters: Encodable {}

File: native/windows/SkillFlow/SkillFlow/Daemon/DaemonClient.cs
using System.Net.Http.Json;
using System.Text.Json;

namespace SkillFlow.Daemon;

public sealed class DaemonClient
{
    public const string ApiVersion = "2026-04-25";

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        PropertyNameCaseInsensitive = true
    };

    private readonly string _endpointPath;
    private readonly HttpClient _httpClient;

    public DaemonClient(HttpClient? httpClient = null, string? endpointPath = null)
    {
        _endpointPath = endpointPath ?? DefaultEndpointPath();
        _httpClient = httpClient ?? new HttpClient();
    }

    public static string DefaultEndpointPath()
    {
        var userProfile = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        return Path.Combine(userProfile, ".skillflow", "runtime", "daemon-service.json");
    }

    public async Task<T> InvokeAsync<T>(
        string method,
        object? parameters = null,
        string? requestId = null,
        CancellationToken cancellationToken = default)
    {
        var endpoint = LoadEndpoint();
        var url = new Uri($"http://{endpoint.Address}/invoke");
        var nativeRequest = new NativeApiRequest(
            ApiVersion,
            method,
            parameters,
            requestId ?? $"win-{Guid.NewGuid():N}");

        var serviceRequest = new DaemonServiceRequest("native.api", nativeRequest);
        using var content = JsonContent.Create(serviceRequest, options: JsonOptions);
        using var request = new HttpRequestMessage(HttpMethod.Post, url)
        {
            Content = content
        };
        request.Headers.Add("X-SkillFlow-Token", endpoint.Token);

        HttpResponseMessage response;
        try
        {
            response = await _httpClient.SendAsync(request, cancellationToken);
        }
        catch (Exception exception) when (exception is HttpRequestException or TaskCanceledException)
        {
            throw new DaemonClientException(
                DaemonClientErrorKind.Transport,
                $"SkillFlow daemon request failed: {exception.Message}");
        }

        await using var disposableResponse = response;
        if ((int)response.StatusCode == 401)
        {
            throw new DaemonClientException(
                DaemonClientErrorKind.Unauthorized,
                "SkillFlow daemon rejected the access token.");
        }

        if (!response.IsSuccessStatusCode)
        {
            throw new DaemonClientException(
                DaemonClientErrorKind.Transport,
                $"SkillFlow daemon returned HTTP {(int)response.StatusCode}.");
        }

        DaemonServiceResponse<T>? serviceResult;
        try
        {
            serviceResult = await response.Content.ReadFromJsonAsync<DaemonServiceResponse<T>>(
                JsonOptions,
                cancellationToken);
        }
        catch (JsonException exception)
        {
            throw new DaemonClientException(
                DaemonClientErrorKind.InvalidResponse,
                $"SkillFlow daemon returned an invalid response: {exception.Message}");
        }

        if (serviceResult is null)
        {
            throw new DaemonClientException(
                DaemonClientErrorKind.InvalidResponse,
                "SkillFlow daemon returned an empty response.");
        }

        if (!serviceResult.Ok)
        {
            throw new DaemonClientException(
                DaemonClientErrorKind.Service,
                serviceResult.Error ?? "SkillFlow daemon service failed.");
        }

        if (serviceResult.Result is null)
        {
            throw new DaemonClientException(
                DaemonClientErrorKind.InvalidResponse,
                "SkillFlow daemon response did not include the native API result.");
        }

        if (!serviceResult.Result.Ok)
        {
            var error = serviceResult.Result.Error;
            throw new DaemonClientException(
                DaemonClientErrorKind.Api,
                error?.Message ?? "The native API request failed.",
                error?.Code ?? "internal_error",
                error?.MessageKey);
        }

        if (serviceResult.Result.Result is null)
        {
            return default!;
        }

        return serviceResult.Result.Result;
    }

    private DaemonEndpoint LoadEndpoint()
    {
        if (!File.Exists(_endpointPath))
        {
            throw new DaemonClientException(
                DaemonClientErrorKind.EndpointUnavailable,
                "SkillFlow daemon endpoint is unavailable.");
        }

        try
        {
            var endpoint = JsonSerializer.Deserialize<DaemonEndpoint>(
                File.ReadAllText(_endpointPath),
                JsonOptions);

            if (endpoint is null ||
                string.IsNullOrWhiteSpace(endpoint.Address) ||
                string.IsNullOrWhiteSpace(endpoint.Token))
            {
                throw new DaemonClientException(
                    DaemonClientErrorKind.InvalidEndpoint,
                    "SkillFlow daemon endpoint is invalid.");
            }

            return endpoint;
        }
        catch (JsonException exception)
        {
            throw new DaemonClientException(
                DaemonClientErrorKind.InvalidEndpoint,
                $"SkillFlow daemon endpoint is invalid: {exception.Message}");
        }
    }
}
